sway/config/output.c

```diff
@@ -0,0 +1,181 @@
+#define _XOPEN_SOURCE 700
+#include <string.h>
+#include <assert.h>
+#include <wlr/types/wlr_output.h>
+#include <wlr/types/wlr_output_layout.h>
+#include "sway/config.h"
+#include "sway/output.h"
+#include "log.h"
+
+int output_name_cmp(const void *item, const void *data) {
+	const struct output_config *output = item;
+	const char *name = data;
+
+	return strcmp(output->name, name);
+}
+
+void output_config_defaults(struct output_config *oc) {
+	oc->enabled = -1;
+	oc->width = oc->height -1;
+	oc->refresh_rate = -1;
+	oc->x = oc->y = -1;
+	oc->scale = -1;
+	oc->transform = -1;
+}
+
+void merge_output_config(struct output_config *dst, struct output_config *src) {
+	if (src->name) {
+		if (dst->name) {
+			free(dst->name);
+		}
+		dst->name = strdup(src->name);
+	}
+	if (src->enabled != -1) {
+		dst->enabled = src->enabled;
+	}
+	if (src->width != -1) {
+		dst->width = src->width;
+	}
+	if (src->height != -1) {
+		dst->height = src->height;
+	}
+	if (src->x != -1) {
+		dst->x = src->x;
+	}
+	if (src->y != -1) {
+		dst->y = src->y;
+	}
+	if (src->scale != -1) {
+		dst->scale = src->scale;
+	}
+	if (src->refresh_rate != -1) {
+		dst->refresh_rate = src->refresh_rate;
+	}
+	if (src->transform != -1) {
+		dst->transform = src->transform;
+	}
+	if (src->background) {
+		if (dst->background) {
+			free(dst->background);
+		}
+		dst->background = strdup(src->background);
+	}
+	if (src->background_option) {
+		if (dst->background_option) {
+			free(dst->background_option);
+		}
+		dst->background_option = strdup(src->background_option);
+	}
+}
+
+static void set_mode(struct wlr_output *output, int width, int height,
+		float refresh_rate) {
+	struct wlr_output_mode *mode, *best = NULL;
+	int mhz = (int)(refresh_rate * 1000);
+	wl_list_for_each(mode, &output->modes, link) {
+		if (mode->width == width && mode->height == height) {
+			if (mode->refresh == mhz) {
+				best = mode;
+				break;
+			}
+			best = mode;
+		}
+	}
+	if (!best) {
+		sway_log(L_ERROR, "Configured mode for %s not available", output->name);
```

Comment:
If no mode is ever set, the drm backend won't ever start up and (I think) the user loses the ability to switch VT so this is a pretty hard crash for a single monitor setup.

The obvious solution is to just set the default mode, but that's a silent error if you don't know to check the logs.

This is an excellent use case for system notifications but we don't have those yet.

Reply:
This is already done here: https://github.com/swaywm/sway/blob/wlroots/sway/desktop/output.c#L111